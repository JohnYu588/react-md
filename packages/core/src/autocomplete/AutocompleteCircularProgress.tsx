import { type ReactElement } from "react";
import {
  CircularProgress,
  type CircularProgressProps,
} from "../progress/CircularProgress.js";
import { type ProgressTheme } from "../progress/types.js";

/**
 * @since 6.0.0
 */
export interface AutocompleteCircularProgressProps
  extends CircularProgressProps {
  /** @defaultValue `"Loading"` */
  "aria-label"?: string;

  /** @defaultValue `"current-color"` */
  theme?: ProgressTheme;
}

/**
 * An internal component used to handle the styling and minimal accessibility
 * for the `CircularProgress` within the `Autocomplete`
 *
 * @internal
 * @since 6.0.0
 */
export function AutocompleteCircularProgress(
  props: AutocompleteCircularProgressProps
): ReactElement {
  const {
    "aria-labelledby": ariaLabelledby,
    "aria-label": ariaLabel = ariaLabelledby ? undefined : "Loading",
    theme = "current-color",
    ...remaining
  } = props;

  return (
    <CircularProgress
      {...remaining}
      aria-label={ariaLabel as string}
      aria-labelledby={ariaLabelledby}
      theme={theme}
    />
  );
}
