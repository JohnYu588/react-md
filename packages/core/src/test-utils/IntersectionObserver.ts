export class IntersectionObserverMock implements IntersectionObserver {
  root: Document | Element | null;
  rootMargin: string;
  thresholds: readonly number[];

  elements: Set<Element>;

  constructor(
    public callback: IntersectionObserverCallback,
    options: IntersectionObserverInit = {}
  ) {
    this.root = options.root || null;
    this.rootMargin = options.rootMargin || "";
    this.thresholds =
      typeof options.threshold === "number"
        ? [options.threshold]
        : options.threshold ?? [];

    this.elements = new Set();
  }

  observe(target: Element): void {
    this.elements.add(target);

    this.callback(this.takeRecords(), this);
  }

  unobserve(target: Element): void {
    this.elements.delete(target);
  }

  takeRecords(): IntersectionObserverEntry[] {
    return [...this.elements].map<IntersectionObserverEntry>((target) => ({
      time: Date.now(),
      target,
      boundingClientRect: target.getBoundingClientRect(),
      intersectionRatio: 0,
      intersectionRect: target.getBoundingClientRect(),
      isIntersecting: false,
      rootBounds:
        this.root && "getBoundingClientRect" in this.root
          ? this.root.getBoundingClientRect()
          : null,
    }));
  }

  disconnect(): void {
    this.elements.clear();
  }
}
