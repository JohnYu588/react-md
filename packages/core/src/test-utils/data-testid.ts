/// <reference types="react" />

declare module "react" {
  interface HTMLAttributes<T>
    extends React.AriaAttributes,
      React.DOMAttributes<T> {
    "data-testid"?: string | number;
  }
}
