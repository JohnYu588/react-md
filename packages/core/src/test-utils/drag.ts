import { fireEvent } from "@testing-library/dom";
import type { MouseEvent } from "react";
import { wait } from "../utils/wait.js";

interface XYCoords {
  x: number;
  y: number;
}

interface BaseDragOptions {
  to?: XYCoords | Element;
  delta?: XYCoords;
  steps?: number;
  duration?: number;
}

type DragOptions = BaseDragOptions &
  ({ to: XYCoords | Element; delta?: never } | { delta: XYCoords; to?: never });

const getElementClientCenter = (element: Element): XYCoords => {
  const { left, top, width, height } = element.getBoundingClientRect();
  return {
    x: left + width / 2,
    y: top + height / 2,
  };
};

const getCoords = (elementOrCoords: Element | XYCoords): XYCoords =>
  "x" in elementOrCoords && "y" in elementOrCoords
    ? elementOrCoords
    : getElementClientCenter(elementOrCoords);

/**
 * This is internal for now. No idea if it's actually worthwhile though since
 * you have to mock the `getBoundingClientRect()` to work.
 *
 * @internal
 * @see https://github.com/testing-library/testing-library-docs/blob/c20f8a8931e9f0a158fda9f01bbea8cb45ed1651/docs/example-drag.mdx
 */
export async function drag(
  element: Element,
  options: DragOptions
): Promise<void> {
  const { delta, to: inTo, steps = 20, duration = 500 } = options;
  const from = getElementClientCenter(element);
  const to = delta
    ? {
        x: from.x + delta.x,
        y: from.y + delta.y,
      }
    : getCoords(inTo);

  const step: XYCoords = {
    x: (to.x - from.x) / steps,
    y: (to.y - from.y) / steps,
  };

  const currentEvent = {
    clientX: from.x,
    clientY: from.y,
    button: 0,
    altKey: false,
    ctrlKey: false,
    metaKey: false,
    shiftKey: false,
  } satisfies Partial<MouseEvent>;

  fireEvent.mouseEnter(element, currentEvent);
  fireEvent.mouseOver(element, currentEvent);
  fireEvent.mouseMove(element, currentEvent);
  fireEvent.mouseDown(element, currentEvent);
  fireEvent.dragStart(element, currentEvent);

  const delay = duration / steps;
  for (let i = 0; i < steps; i++) {
    currentEvent.clientX += step.x;
    currentEvent.clientY += step.y;
    await wait(delay);
    fireEvent.mouseMove(element, currentEvent);
  }

  fireEvent.mouseUp(element, currentEvent);
  fireEvent.dragEnd(element, currentEvent);
}
