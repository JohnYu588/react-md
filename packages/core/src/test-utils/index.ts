export * from "@testing-library/react";
export * from "@testing-library/user-event";
export * from "./mocks/match-media.js";
export * from "./render.js";
export * from "./utils/resize-observer.js";
