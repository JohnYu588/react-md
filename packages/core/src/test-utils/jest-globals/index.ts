export * from "./match-media.js";
export * from "./resize-observer.js";
export * from "./timers.js";
