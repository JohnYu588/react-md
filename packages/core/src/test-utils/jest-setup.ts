import { beforeEach } from "@jest/globals";
import { INTERACTION_CONFIG } from "../interaction/config.js";

beforeEach(() => {
  // set the mode to `none` in tests since ripples require
  // `getBoundingClientRect()` to create correct CSS. You'll either see warnings
  // in the console around invalid css values or `NaN`.
  INTERACTION_CONFIG.mode = "none";
});
