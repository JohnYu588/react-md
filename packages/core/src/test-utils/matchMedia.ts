import { jest } from "@jest/globals";
import { act } from "@testing-library/react";
import {
  DEFAULT_DESKTOP_LARGE_MIN_WIDTH,
  DEFAULT_DESKTOP_MIN_WIDTH,
  DEFAULT_PHONE_MAX_WIDTH,
  DEFAULT_TABLET_MIN_WIDTH,
} from "../media-queries/appSize.js";

const noop = (): void => {
  // do nothing
};

/**
 * @see {@link spyOnMatchMedia} instead
 * @internal
 * @remarks \@since 6.0.0
 */
export const BASE_MEDIA_QUERY_LIST: MediaQueryList = {
  media: "",
  matches: false,
  onchange: noop,
  addListener: noop,
  addEventListener: noop,
  removeEventListener: noop,
  removeListener: noop,
  dispatchEvent: () => false,
};

/** @remarks \@since 6.0.0 */
export type MatchMediaMatcher = (query: string) => boolean;

/**
 * @see {@link spyOnMatchMedia} for usage
 * @remarks \@since 6.0.0
 * @returns `true` for phone media queries
 */
export const matchPhone: MatchMediaMatcher = (query) =>
  query.includes(`${DEFAULT_PHONE_MAX_WIDTH}`);

/**
 * @see {@link spyOnMatchMedia} for usage
 * @remarks \@since 6.0.0
 * @returns `true` for tablet media queries
 */
export const matchTablet: MatchMediaMatcher = (query) =>
  query.includes(`${DEFAULT_TABLET_MIN_WIDTH}`);

/**
 * @see {@link spyOnMatchMedia} for usage
 * @remarks \@since 6.0.0
 * @returns `true` for desktop media queries
 */
export const matchDesktop: MatchMediaMatcher = (query) =>
  query.includes(`${DEFAULT_DESKTOP_MIN_WIDTH}`);

/**
 * @see {@link spyOnMatchMedia} for usage
 * @remarks \@since 6.0.0
 * @returns `true` for large desktop media queries
 */
export const matchLargeDesktop: MatchMediaMatcher = (query) =>
  query.includes(`${DEFAULT_DESKTOP_LARGE_MIN_WIDTH}`);

/**
 * @see {@link spyOnMatchMedia} for usage
 * @remarks \@since 6.0.0
 * @returns `true` for both desktop and large desktop media queries
 */
export const matchAnyDesktop: MatchMediaMatcher = (query) =>
  matchDesktop(query) || matchLargeDesktop(query);

/**
 * @remarks \@since 6.0.0
 */
export type MatchMediaSpiedFunction = jest.SpiedFunction<
  typeof window.matchMedia
> & {
  /**
   * @example
   * Default Behavior
   * ```tsx
   * const matchMedia = spyOnMatchMedia();
   * render(<Test />);
   *
   * // expect desktop results
   *
   * matchMedia.changeViewport(matchPhone);
   * // expect phone results
   * ```
   *
   * @example
   * Custom Act Behavior
   * ```tsx
   * const matchMedia = spyOnMatchMedia();
   * const { rerender } = render(<Test />);
   *
   * // expect desktop results
   *
   * matchMedia.changeViewport(matchPhone, false);
   * rerender(<Test key="new-key" />);
   *
   * // expect phone results
   * ```
   */
  changeViewport(matcher: MatchMediaMatcher, disableAct?: boolean): void;
};

/**
 * @example
 * Default Behavior
 * ```tsx
 * import { matchPhone, render, spyOnMatchMedia } from "@react-md/test-utils";
 *
 * const matchMedia = spyOnMatchMedia();
 * render(<Test />);
 *
 * // expect desktop results
 *
 * matchMedia.changeViewport(matchPhone);
 * // expect phone results
 * ```
 *
 * @example
 * Set Default Media
 * ```tsx
 * import { matchPhone, render, spyOnMatchMedia } from "@react-md/test-utils";
 *
 * const matchMedia = spyOnMatchMedia(matchPhone);
 * render(<Test />);
 *
 * // expect phone results
 * ```
 *
 * @remarks \@since 6.0.0
 */
export function spyOnMatchMedia(
  defaultMatch: MatchMediaMatcher = matchDesktop
): MatchMediaSpiedFunction {
  type Listener = (event: MediaQueryListEvent) => void;

  const listeners = new Map<string, Listener>();
  const matchMedia = jest
    .spyOn(window, "matchMedia")
    .mockImplementation((query) => ({
      ...BASE_MEDIA_QUERY_LIST,
      addEventListener(type: string, listener: Listener | EventListenerObject) {
        /* c8 ignore start */
        if (typeof listener !== "function" || type !== "change") {
          return;
        }
        /* c8 ignore end */

        listeners.set(query, listener);
      },
      removeEventListener(
        type: string,
        listener: Listener | EventListenerObject
      ) {
        /* c8 ignore start */
        if (typeof listener !== "function" || type !== "change") {
          return;
        }
        /* c8 ignore end */

        listeners.delete(query);
      },
      matches: defaultMatch(query),
    }));

  const changeViewport = (
    matcher: MatchMediaMatcher,
    disableAct = false
  ): void => {
    const update = (): void => {
      window.dispatchEvent(new Event("resize"));

      const event = new Event("change");
      listeners.forEach((listener, query) => {
        listener({
          ...event,
          media: "",
          matches: matcher(query),
        });
      });
    };
    if (disableAct) {
      update();
    } else {
      act(update);
    }
  };

  const mock = matchMedia as MatchMediaSpiedFunction;
  mock.changeViewport = changeViewport;

  return mock;
}
