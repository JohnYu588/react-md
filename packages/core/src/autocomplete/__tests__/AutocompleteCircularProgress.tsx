import { describe, expect, it } from "@jest/globals";
import { render, screen } from "test-utils";
import { AutocompleteCircularProgress } from "../AutocompleteCircularProgress.js";

describe("AutocompleteCircularProgress", () => {
  it("should apply the correct styling and HTML attributes", () => {
    const { rerender } = render(<AutocompleteCircularProgress />);

    const progress = screen.getByRole("progressbar", { name: "Loading" });
    expect(progress).toMatchSnapshot();

    rerender(
      <AutocompleteCircularProgress
        style={{ color: "white" }}
        className="custom-class-name"
      />
    );
    expect(progress).toMatchSnapshot();
  });

  it("should allow for a custom aria-label", () => {
    const { rerender } = render(<AutocompleteCircularProgress />);

    const progress = screen.getByRole("progressbar", { name: "Loading" });
    expect(progress).toHaveAttribute("aria-label", "Loading");
    expect(progress).not.toHaveAttribute("aria-labelledby");

    rerender(<AutocompleteCircularProgress aria-label="Custom" />);
    expect(progress).toHaveAttribute("aria-label", "Custom");
    expect(progress).not.toHaveAttribute("aria-labelledby");
  });

  it("should not default the aria-label if aria-labelledby was provided", () => {
    render(<AutocompleteCircularProgress aria-labelledby="some-other-id" />);
    const progress = screen.getByRole("progressbar");
    expect(progress).not.toHaveAttribute("aria-label");
    expect(progress).toHaveAttribute("aria-labelledby", "some-other-id");
  });
});
