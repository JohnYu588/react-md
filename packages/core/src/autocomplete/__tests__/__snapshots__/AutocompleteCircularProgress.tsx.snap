// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutocompleteCircularProgress should apply the correct styling and HTML attributes 1`] = `
<span
  aria-label="Loading"
  aria-valuemax="100"
  aria-valuemin="0"
  class="rmd-circular-progress rmd-circular-progress--centered"
  id="circular-progress-:r0:"
  role="progressbar"
>
  <svg
    class="rmd-circular-progress__svg rmd-circular-progress__svg--indeterminate"
    viewBox="0 0 66 66"
  >
    <circle
      class="rmd-circular-progress__circle rmd-circular-progress__circle--indeterminate"
      cx="33"
      cy="33"
      r="30"
    />
  </svg>
</span>
`;

exports[`AutocompleteCircularProgress should apply the correct styling and HTML attributes 2`] = `
<span
  aria-label="Loading"
  aria-valuemax="100"
  aria-valuemin="0"
  class="rmd-circular-progress rmd-circular-progress--centered custom-class-name"
  id="circular-progress-:r0:"
  role="progressbar"
  style="color: white;"
>
  <svg
    class="rmd-circular-progress__svg rmd-circular-progress__svg--indeterminate"
    viewBox="0 0 66 66"
  >
    <circle
      class="rmd-circular-progress__circle rmd-circular-progress__circle--indeterminate"
      cx="33"
      cy="33"
      r="30"
    />
  </svg>
</span>
`;
